hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java

```diff
@@ -4074,29 +4183,128 @@ private static long getLongValue(final Cell cell) throws DoNotRetryIOException {
       return PrivateCellUtil.getValueAsLong(cell);
     }
 
-    /**
-     * Do a specific Get on passed <code>columnFamily</code> and column qualifiers.
-     * @param mutation Mutation we are doing this Get for.
-     * @param store Which column family on row (TODO: Go all Gets in one go)
-     * @param coordinates Cells from <code>mutation</code> used as coordinates applied to Get.
-     * @return Return list of Cells found.
-     */
-    private List<Cell> get(Mutation mutation, HStore store, List<Cell> coordinates,
-      TimeRange tr) throws IOException {
-      // Sort the cells so that they match the order that they appear in the Get results.
-      // Otherwise, we won't be able to find the existing values if the cells are not specified
-      // in order by the client since cells are in an array list.
-      // TODO: I don't get why we are sorting. St.Ack 20150107
```

Comment:
Anything on this sort issue?

Reply:
We need the sorting actually, otherwise it will return a wrong result to the client. I keep this sorting here: https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L4072-L4075

Reply:
Let me explain this.

There was one requirement where the additional cells of a mutation need to be visible for followed mutations. For example, when we have a Put operation and a CheckAndMutate operation in a single mutation batch, the additional cells of the Put operation should be visible for the followed CheckAndMutate operation.

The following code keeps additional cells of mutations:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L3833-L3835

To achieve this, I first tried to use the existing `additionalScanners` of `RegionScannerImpl`:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L3061

However, I found it's much easier to add a new `additionalScanners` to `StoreScanner` and use it than to use the existing `additionalScanners` of `RegionScannerImpl`.

So, I decided to add `additionalScanners` to `StoreScanner` and was able to make the additional cell visible for the followed mutations.
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java#L251-L253

By the way, the test code of this is as as follows:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java#L7101-L7174

Thanks.